apps/extension/src/storage/base.ts

```diff
@@ -20,30 +20,53 @@ export interface StorageItem<T> {
 }
 
 type Version = string;
+
+export type MigrationMap<OldState, NewState> = {
+  [K in keyof OldState & keyof NewState]?: (
+    prev: OldState[K],
+    get: <K extends keyof NewState>(key: K) => Promise<NewState[K]>,
+  ) => NewState[K] | Promise<NewState[K]>;
+};
+
 // It is quite difficult writing a generic that covers all migration function kinds.
 // Therefore, the writer of the migration should ensure it is typesafe when they define it.
-// See `migration.test.ts` for an example.
 // eslint-disable-next-line @typescript-eslint/no-explicit-any
-type Migration = Record<Version, (a: any) => any>;
-type Migrations<K extends string | number | symbol> = Partial<Record<K, Migration>>;
+export type Migrations<T> = Partial<Record<Version, MigrationMap<any, T>>>;
+
+export type VersionSteps = Record<Version, Version>;
 
 export class ExtensionStorage<T> {
   constructor(
     private storage: IStorage,
     private defaults: T,
     private version: Version,
-    private migrations: Migrations<keyof T>,
+    private migrations: Migrations<T> = {},
+    private versionSteps: VersionSteps = {},
   ) {}
 
   async get<K extends keyof T>(key: K): Promise<T[K]> {
+    return await this.getRaw({ key, skipMigration: false });
+  }
+
+  private async getRaw<K extends keyof T>({
+    key,
+    skipMigration,
+  }: {
+    key: K;
+    skipMigration: boolean;
+  }): Promise<T[K]> {
     const result = (await this.storage.get(String(key))) as
       | Record<K, StorageItem<T[K]>>
       | EmptyObject;
 
     if (isEmptyObj(result)) {
       return this.defaults[key];
     } else {
-      return await this.migrateIfNeeded(key, result[key]);
+      if (skipMigration) {
+        return result[key].value;
+      } else {
+        return await this.migrateIfNeeded(key, result[key]);
+      }
```

Comment:
Given some migrations need access to the wider storage getter, we have two ways to get from storage now.

Reply:
resolving as outdated